framework/core/instance.cpp

```diff
@@ -381,20 +381,54 @@ void Instance::query_gpus()
 
 PhysicalDevice &Instance::get_suitable_gpu()
 {
-	assert(!gpus.empty() && "No physical devices were found on the system.");
+	if (gpus.size() == 0)
+	{
+		throw std::runtime_error("No physical devices found.");
+	}
+
+	uint32_t count_device_type[VK_PHYSICAL_DEVICE_TYPE_CPU + 1];
+	memset(count_device_type, 0, sizeof(count_device_type));
+
+	for (uint32_t i = 0; i < gpus.size(); i++)
+	{
+		const auto deviceType = gpus[i]->get_properties().deviceType;
+		assert(deviceType <= VK_PHYSICAL_DEVICE_TYPE_CPU);
+		count_device_type[deviceType]++;
+	}
+
+	const VkPhysicalDeviceType device_type_preference[] = {
```

Comment:
The project uses `std::array` and `std::vector` throughout the code. Not sure why you went with raw arrays and C style `sizeof(device_type_preference) / sizeof(VkPhysicalDeviceType)`?